SPRacingF7DUAL - Dual SIMULTANEOUS gyro support.

hydra opened the pull request:
The dual simultaneous gyro support is work of @martinbudden and @hydra

SPRacingF7DUAL has BOTH the gyros on the same SPI bus, this required a few changes to the gyro bus and SPI pin initialization order, especially when only one gyro is being used. (i.e. you still need to initialize the CS pin on the gyro you're not using...)

@DieHertz I'll address your PR comments from the CF PR here since the branch on CF for the SPRacingF7DUAL is not `master`.

Ported from Cleanflight - https://github.com/cleanflight/cleanflight/pull/3016/commits

Status
- [x] Test flown.
- [x] Test other dual gyro boards which have gyros on difference SPI instances (OmnibusF7, SPRacingF4EVODG).

DieHertz commented:
Tested on the bench, also works while running from ITCM-FLASH interface with performance figures similar to NERO.
DSHOT1200+16/16+dynamic filter at 24%.

For some reason, reboot to bootloader doesn't work neither from CLI nor from Configurator Firmware Flashing tab.

Have you noticed some warnings regarding `accgyro.h`?
```
./src/main/drivers/accgyro/accgyro.h:72:24: warning: padding struct to align 'mpuConfiguration' [-Wpadded]
./src/main/drivers/accgyro/accgyro.h:70:11: warning: padding struct to align 'gyroADCf' [-Wpadded]
```

mikeller commented:
@hydra: This combines a lot of different things into one pull request: Dual gyro / VTX updates / button updates / 2 new targets / ... In order to get this merged, and deal with potential follow up to make it work for other targets, I think it should be broken down into one pull request per changed area / target. Thanks.

DieHertz commented:
At this point it was really helpful to get my SPRF7DUAL up & running with BF for F7 performance fixes.
For 3.4 I think we should extract linker scripts changes to a different PR, as well as other sub-features, culminating with a PR which builds on all previous PRs and adds the SPRF7DUAL target.

hydra commented:
@mikeller agreed. I wanted to get your eyes on it and thoughts and also so that @diehertz had a working target.  I'll extract the code into a set of smaller PRs ASAP.

mikeller commented:
I had a look at the code, and it looks good to me. The risk of the changes - in particularly the gyro changes - seems to be relatively high, so I think this should go into 3.4. What do you think of extracting a pull request for a `SPRACINGF7SINGLE` (or just `SPRACINGF7`), to go into 3.3?

jflyper commented:
I think creating `SPRACINGF7SINGLE` is a bad move.

mikeller commented:
@jflyper: Better than not having a target for the hardware in 3.3 at all. It could also be an early version of `SPRACINGF7DUAL`, but the name would be somewhat misleading.

hydra commented:
yes, creating SPRacingF7 without the dual gyro is a bad idea, i tried yesterday and it's more problematic and more effort than just reviewing the dual gyro changes due to the initialisation and detection of the gyro and the fact the second gyro must be configured still otherwise the first gyro will not function.

mikeller commented:
Ah, that's a different situation to the other dual gyro board then (`OMNIBUSF4`?).

hydra commented:
@mikeller theres not much risk tbh, mostly its just ensuring that all the pins for the gyros are always configured correctly.  the dual gyro sensor reading is also trivial, there's also more debugging/logging code than flight code changes.

I've been flying and testing a variety of targets with/without dual gyro in CF since may last year so the init code changes are pretty well tested.

hydra commented:
yes, ALL the other dual gyro targets use one bus per sensor.

mikeller commented:
@hydra: No problem, should be safe to be added for 3.4 then.

hydra commented:
I'll do some PRs later on today, working on the house at the moment.

hydra commented:
@mikeller OK, I've extracted 7 PR's from this PR.

https://github.com/betaflight/betaflight/pull/5274
https://github.com/betaflight/betaflight/pull/5276
https://github.com/betaflight/betaflight/pull/5278
https://github.com/betaflight/betaflight/pull/5279
https://github.com/betaflight/betaflight/pull/5280
https://github.com/betaflight/betaflight/pull/5281
https://github.com/betaflight/betaflight/pull/5282

I rebased this on master, keeping only the ESSENTIAL commits that are also in the other PR's, I'll rebase this branch as they are merged.

I extracted the SPRacingF4EVODG target to a new commit, it's still in this PR but requires other changes in this PR to build.

Hope you can have a look at the other PR's and get them merged for 3.3 as there's a good few cleanups and fixes in them.

DieHertz commented:
Sweet, GJ Dominic

hydra commented:
I'll get this PR rebased now some of the other PRs are merged.

hydra commented:
rebased

hydra commented:
tasks in description completed (test flown, tested other dual gyro target that has two gyros on different SPI bus instances)

hydra commented:
builds are failing due to flash space issues since rebase.  See #5323 which might help a little

mikeller commented on src/main/sensors/gyro.c:
```diff
@@ -454,8 +478,80 @@ bool gyroInit(void)
+    gyroSensor1.gyroDev.gyroAlign = GYRO_1_ALIGN;
+#endif
+    gyroSensor1.gyroDev.bus.bustype = BUSTYPE_SPI;
+    spiBusSetInstance(&gyroSensor1.gyroDev.bus, GYRO_1_SPI_INSTANCE);
```
This is causing a compilation error in KAKUTEF4.

hydra replied:
fixed.

hydra commented:
rebased and removed the #5323 commit for "conditional debug names" that this PR also included.

hydra commented:
builds failing due to flash size issues, https://github.com/betaflight/betaflight/pull/5287 'optimize math' will help here.

SRWieZ commented:
Hi ! 

I'm a bit curious, I found this interesting so I have some questions.
I'm a web developer for so long that I'm not sure if I read it right 

Actually is it just gyro1+gyro2/2 ?
Did you see any improvement in flight ? 
What kind of filter do we need to run with dual gyro ?
If i read it right, we can't do different filter on different gyro. Is it planned ? I don't even know if it's a good idea, just asking.

Thanks!

mikeller commented on src/main/target/SPRACINGF4EVO/target.h:
```diff
@@ -34,22 +37,40 @@
 
-#define USE_MPU_DATA_READY_SIGNAL
-#define ENSURE_MPU_DATA_READY_IS_LOW
+//#define USE_MPU_DATA_READY_SIGNAL
```
Can probably be removed.

hydra replied:
removed for the EVODG target, kept for EVO target - via `#ifdef`

mikeller commented on src/main/target/SPRACINGF7DUAL/config.c:
```diff
@@ -0,0 +1,67 @@
+void targetConfiguration(void)
+{
+    barometerConfigMutable()->baro_hardware = BARO_DEFAULT;
+#if (0) // XXX remove this check, for debugging serial port conflicts
```
Should this be in here?

hydra replied:
removed

mikeller commented on src/main/target/SPRACINGF7DUAL/sound.c:
```diff
@@ -0,0 +1,152 @@
+    HAL_DAC_Start(&hdac, DAC_CHANNEL_1);
+
+
+#if 0
```
If you want to define out code sections, then please use a name, so that users can enable the code on the command line when building.

hydra replied:
yup, this will all be sorted out in the audio PR.

hydra replied:
actually, I'll just delete the sound code from this PR for now.

mikeller commented on src/main/target/SPRACINGF7DUAL/sound.c:
```diff
@@ -0,0 +1,152 @@
+/*
```
I think this should go into `src/main/drivers`.

hydra replied:
I have a new feature that will do exactly that.  new feature in new PR, this just lays the groundwork for experimenters.

jflyper commented on src/main/drivers/bus_spi.h:
```diff
@@ -56,7 +56,7 @@ typedef enum {
     SPI_CLOCK_SLOW          = 256, //00.42188 MHz
     SPI_CLOCK_STANDARD      = 16,  //06.57500 MHz
-    SPI_CLOCK_FAST          = 4,   //27.00000 MHz
+    SPI_CLOCK_FAST          = 8,   //13.50000 MHz
```
This needs some consideration.

Is `SPI_CLOCK_FAST` only used for gyros atm?

hydra replied:
as per slack conversation, no SD card, flash etc.

hydra commented:
@SRWieZ Hi! thanks for your interest.

There's more to it on a physical level, the alignment of each sensor is different therefore vibrations affect each sensor differently, by combining the two gyros outputs you get a system that's more accurate.  And yes it flies great, I was amazed how smooth it was before even thinking about filter configuration.

It would be totally possible to apply different filters to each gyro, if you have time to work on that then please go right ahead!

jflyper commented on src/main/fc/fc_init.c:
```diff
@@ -196,52 +197,58 @@ static IO_t busSwitchResetPin        = IO_NONE;
     spiPreInitCs(IO_TAG(MPU6000_CS_PIN));
 #endif
-#ifdef USE_ACC_SPI_MPU6500
+#ifdef MPU6500_CS_PIN
```
I think this is not right.

USE and PIN are different.

hydra replied:
yes, 100% agreed. however if a target defines the pin then it MUST be initialised, even if the sensor detection code is compiled out.

jflyper replied:
I think PIN were converted to USE at some point in the past, to distinguish USE and PIN. I forgot what the reason was, though. May be a provisioning for drivers that require additional conditional inside the USE, or handling multiple pins.

jflyper replied:
Conversation at slack:
> basically, if you don't initialise all CS pins then the device, if it's unused but present on the PCB will be ACTIVE and start responding to data on the SPI MOSI/MISO lines that is meant for another device.

I'm okay with the change now.

hydra commented:
rebased

hydra commented:
🎉 🎉 🎉
